src/eos-updater-poll-lan.c

```diff
@@ -0,0 +1,920 @@
+
+struct _EosServiceWithBranchFile
+{
+  GObject parent_instance;
+
+  EosAvahiService *service;
+  EosBranchFile *branch_file;
+
+  GDateTime *declared_download_time;
+  gchar *declared_sha512sum;
+};
+
+static void
+eos_service_with_branch_file_dispose_impl (EosServiceWithBranchFile *swbf)
+{
+  g_clear_object (&swbf->service);
+  g_clear_object (&swbf->branch_file);
+  g_clear_pointer (&swbf->declared_download_time, g_date_time_unref);
+}
+
+static void
+eos_service_with_branch_file_finalize_impl (EosServiceWithBranchFile *swbf)
+{
+  g_free (swbf->declared_sha512sum);
+}
+
+EOS_DEFINE_REFCOUNTED (EOS_SERVICE_WITH_BRANCH_FILE,
+                       EosServiceWithBranchFile,
+                       eos_service_with_branch_file,
+                       eos_service_with_branch_file_dispose_impl,
+                       eos_service_with_branch_file_finalize_impl)
+
+static EosServiceWithBranchFile *
+eos_service_with_branch_file_new (EosAvahiService *service)
+{
+  EosServiceWithBranchFile *swbf = g_object_new (EOS_TYPE_SERVICE_WITH_BRANCH_FILE, NULL);
+
+  swbf->service = g_object_ref (service);
+  swbf->branch_file = eos_branch_file_new_empty ();
+
+  return swbf;
+}
+
+static void
+eos_service_with_branch_file_get_from_lan_data (EosServiceWithBranchFile *swbf,
+                                                LanData *lan_data)
+{
+  g_set_object (&swbf->branch_file, lan_data->fetch_data->data->branch_file);
+}
+
+typedef gboolean (*ServiceCheck) (LanData *lan_data,
+                                  EosServiceWithBranchFile *swbf,
+                                  gboolean *valid,
+                                  GError **error);
+
+typedef enum
+  {
+    SERVICE_INVALID,
+    SERVICE_VALID_MORE_CHECKS,
+    SERVICE_VALID
+  } ServiceValidResult;
+
+static gboolean
+check_ostree_path (LanData *lan_data,
+                   const gchar *ostree_path,
+                   gboolean *ok,
+                   GError **error)
+{
+  g_autofree gchar *current_ostree_path = NULL;
+  OstreeRepo *repo = lan_data->fetch_data->data->repo;
+
+  if (!eos_updater_get_ostree_path (repo, &current_ostree_path, error))
+    return FALSE;
+
+  *ok = (g_strcmp0 (ostree_path, current_ostree_path) == 0);
+  return TRUE;
+}
+
+typedef enum
+  {
+    DL_TIME_NEW,
+    DL_TIME_OLD,
+    DL_TIME_INVALID,
+    DL_TIME_OUT_OF_RANGE,
+  } DlTimeCheckResult;
+
+static DlTimeCheckResult
+check_dl_time (LanData *lan_data,
+               const gchar *dl_time,
+               GDateTime **utc_out)
+{
+  gint64 utc_time;
+  gchar *utc_str_end = NULL;
+  g_autoptr(GDateTime) utc = NULL;
+
+  if (dl_time[0] == '\0')
+    return DL_TIME_INVALID;
+  errno = 0;
+  utc_time = g_ascii_strtoll (dl_time, &utc_str_end, 10);
+  if (errno == EINVAL)
+    return DL_TIME_INVALID;
+  if (errno == ERANGE)
+    return DL_TIME_OUT_OF_RANGE;
+  if (*utc_str_end != '\0')
+    return DL_TIME_INVALID;
+  utc = g_date_time_new_from_unix_utc (utc_time);
+  if (utc == NULL)
+    return DL_TIME_INVALID;
```

Comment:
The documentation says that `g_date_time_new_from_unix_utc` returns `NULL` when it is out of range: https://developer.gnome.org/glib/stable/glib-GDateTime.html, so maybe it might be better to return `DL_TIME_OUT_OF_RANGE`.